### PR TITLE
Dynamic Event Details screen

### PR DESCRIPTION
**Pull request by Viccha18**
# Objective

Link the Joyti's viewModel for events details with the Event Details screen made in sprint 3 

# Detailed changes
**Event Details Screen UI**
- Made a separate file for the event UI component (will add in the futur the "editable" components for Create Event screen)
- Added Constraint Layout
- Removed static string in Text(), and replaced them with ViewModel EventUiState
- Added many string ressource used in the display of an event (Title words such as "description" "date"... & conversion for the EventCategory enum)

**Navigation**
- Added composable route to navGraph and take a "eventId" argument
- call navigate with Route.EVENT_DETAILS & eventId from Home.kt 

**EventCategory enum**
- Added a `toString(context: Context)` fonction to Event Category enum to get a res defined string to display
- Added "PARTY" ~and "UNDEFINED"~ in EventCategory enum

**EventDetailsViewModel**
- Removed initialzation function of the EventUiState dataclass which was not necessary
- Changed var to val in the EventUiState dataclass 
- ~Create instance of Event from fetched data. Was: (data from firebase --> EventUiState) Now: (data from firebase --> new Event --> EventUiState)~

**Tests**
- Created a mockViewModel in for the existing display test, and a sample EventUiState
- one test for EventDetailsViewModel

# Note 
~- TODO need to mock the firestore to test the viewModel in depth. 
Christine and I are thinking about making a "Database" class (like proposed in the bootcamp) that would contain the function that interact with the database and that takes care of converting the data from the firebase format to our own data model (Event, User, ....)~

### CODE REVIEW

File: app/src/main/java/com/github/se/eventradar/model/event/Event.kt
Comment (jgeaso1266) [inline app/src/main/java/com/github/se/eventradar/model/event/Event.kt:15]: These my clash with #125

**Review by jgeaso1266 — CHANGES_REQUESTED** [review]
Wow, big PR. A few important things that should be addressed, but mostly comments and questions.

File: app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt:51]: Since I separted the "description "distance" "date & time" etc.. in different composable function I wanted to avoid
- having a tons of function parameters
- to hard code the color scheme in each component when they need to be harmonized

File: app/src/main/java/com/github/se/eventradar/ui/event/EventUiComponents.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/ui/event/EventUiComponents.kt:77]: As we discussed, when a user browse events and clicks on the details, he don't see the location but at what distance the event is from him. Currently, the Location object contains geoPoint and address, so we still to have a function that takes those geo points and the position of the user to compute the distance between them

File: app/src/main/java/com/github/se/eventradar/model/event/Event.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/Event.kt:17]: ok, I'll make some more research, or do you know how I can get it ? I tried stringResource() but it needed a composable function.

File: app/src/main/java/com/github/se/eventradar/model/event/EventDetailsViewModel.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/EventDetailsViewModel.kt:107]: Sorry, I should have read my comments before pushing. I meant a default value. One that we can use by default if there's problem with the database data (similar to the UNDEFINED EventCategory that I added in case the "category field" doesn't contain a valide category)

File: app/src/main/java/com/github/se/eventradar/model/event/EventDetailsViewModel.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/EventDetailsViewModel.kt:64]: That's a good point. I like to separate things (maybe to much). I am wondering what the Event model will be used for. Because one of my next task is to create the Create event screen and add functions to the viewModel, and naturally later, there will be the Edit event Screen as well...

You're right that there is no need now to have them separate, so I can remove that middle man.

File: app/src/main/java/com/github/se/eventradar/model/event/Event.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/Event.kt:15]: Right ty, I will discuss it with Christine

**Review by RaulinN — COMMENTED** [review]
Couldn't review the 100% thoroughly because I couldn't find a way to run the code on the emulator (didn't have any eventIds and clicking on an event from the homepage did not seem to work. Overall, from what I could see, that's some solid work well done!

**Comment by wisdomseeker2001** [comment]
Hello,

I took over Vicky's Branch (https://github.com/swent-sp-2024-event-radar/event-radar-app/pull/126) and made a duplicate. It needed to be adapted to the new structure of the project, that being mainly a new class for all the firebase communication (FirebaseEventRepository.kt) and also the wonderful modifications that @jgeaso1266 made the Event.kt, making the code a lot more clean.

I only changed/ added two classes:

Modified EventDetailsViewModel.kt:
-Adapted it such that it calls the getEvent() function in FirebaseEventRepository.kt and then updated the Ui State.

Created EventOverviewViewmodelUnitTest.kt:
This is a really basic test just to see if the whole object was initialised correctly,
If it fails, I will then break down to assert each of the objects parameter's to see which one is not being initialised properly

I also added the sample event to the Firebase :)

all the other changes you see are not mine, I couldn't get them to go away... they have all even commented on in https://github.com/swent-sp-2024-event-radar/event-radar-app/pull/126

Please let me know if you see any errors

File: app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt
Comment (jgeaso1266) [inline app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt:15]: this is already implemented by [Kotlin](https://stackoverflow.com/questions/28548015/how-do-i-create-an-enum-from-a-string-in-kotlin)

File: app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt
Comment (jgeaso1266) [inline app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt:49]: Consider putting this in the `init` block of the view model instead

File: app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt
Comment (jgeaso1266) [reply app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt:56]: Ok, make sure this is in line with what the rest of the app looks like.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt
Comment (jgeaso1266) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt:21]: These shouldn't be deleted...

**Review by jgeaso1266 — COMMENTED** [review]
Good overall but a few questions and suggestions.

File: app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt:15]: I know there already exist a function from kotlin that will convert the enum value to string, but the idea (maybe I think touo far) is to map each type of the enum to a Ressource String, which as I understood is of use to easily translate the app and its content. Since I used string ressource for subtitle such as "description", "location" I wanted to stay consistent and have string ressource to display the event Category as well

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt
Comment (Nath-Assa) [inline app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt:52]:
```suggestion
       is Resource.Failure -> Log.e("EventDetailsViewModel", "Error getting event: ${response.throwable.message}")
```

File: app/src/test/java/com/github/se/eventradar/EventDetailsViewmodelUnitTest.kt
Comment (Nath-Assa) [inline app/src/test/java/com/github/se/eventradar/EventDetailsViewmodelUnitTest.kt:1]: Consider expanding the test coverage for EventDetailsViewModel to account for a variety of scenarios. To begin, you can verify that all properties of EventUiState are correctly updated after data retrieval. Additionally, it would be beneficial to implement tests that examine how the ViewModel handles scenarios where the event data is modified or removed from the database, ensuring the model reacts appropriately to data changes or unavailability.

**Review by Nath-Assa — COMMENTED** [review]
Nice work! I made some suggestions. I would also like you to add a few more tests to the ViewModel before I approve.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt
Comment (Nath-Assa) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt:21]: If the `toString(context: Context)` method in the EventCategory enum simplifies the usage of localized string resources, I recommend keeping it. However, to enhance readability and avoid confusion with the standard toString() method, consider renaming it to something more descriptive like `getLocalizedName(context: Context)`

File: app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt
Comment (jgeaso1266) [inline app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt:16]: Still don't think this is necessary. It adds more complexity to the implementation when one can just use the enum.

File: app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt
Comment (jgeaso1266) [reply app/src/main/java/com/github/se/eventradar/ui/event/EventDetails.kt:57]: Is the color the same as the rest of the app? I don't think we have implemented specific colors like you have (we are currently just using default colors). If that is the case, then maybe take this out of this PR for now so it can be uniform for M2.

File: app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt
Comment (jgeaso1266) [reply app/src/main/java/com/github/se/eventradar/viewmodel/EventDetailsViewModel.kt:21]: > `initialEventState()` which was used to initialized the UiState does not exist anymore (initialization is now done directly in the dataclass) maybe you're talking about
> 
> ```kotlin
>   private val _uiState = MutableStateFlow(EventUiState())
>   val uiState: StateFlow<EventUiState> = _uiState
> ```
> 
> at line 31&32

I didn't see this. For visibility and consistency with other view models, this should be the first two lines of the view model.

**Review by jgeaso1266 — COMMENTED** [review]
A few questions. I am not sure if they require fixes so not going to "request changes" but will approve once you answer the questions.

File: app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt
Comment (Viccha18) [reply app/src/main/java/com/github/se/eventradar/model/event/EventCategory.kt:16]: I cannot reply under your other comments so I'll answer here

1. I can agree that it is of no use in V1 of our app. But as a french speaker, I would like to have the app in french lol, and we wouldn't put every translation in the enum. This is why I implemented this using the resource string. But if you're strongly against having that function in V1, I can remove it.

2. Colors are exactly the same as the one already used in the rest of the app. I use the default colorScheme with "onSurface" "onSurface Variant" etc... I just didn't hard coded the color scheme in my individual UI components. We can discuss the relevance of this. Because at that point in the app, colorScheme colors could be hard coded in the components since they always take the same primary material and we have currently no use to switch them to secondary color scheme.

**Review by jgeaso1266 — APPROVED** [review]
LGTM

**Pull request merged**
